components/admin_console/permission_schemes_settings/guest_permissions_tree/guest_permissions_tree.tsx

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {useCallback, useMemo} from 'react';
+import {FormattedMessage} from 'react-intl';
+
+import Permissions from 'mattermost-redux/constants/permissions';
+import {Role} from 'mattermost-redux/types/roles';
+import {ClientLicense} from 'mattermost-redux/types/config';
+
+import PermissionGroup from '../permission_group.jsx';
+import EditPostTimeLimitButton from '../edit_post_time_limit_button';
+import EditPostTimeLimitModal from '../edit_post_time_limit_modal';
+
+type Props = {
+    license: ClientLicense;
+    onToggle: (role: string, permissions: string[]) => void;
+    readOnly: boolean;
+    scope: string;
+    selectRow: (permission: string) => void;
+    parentRole?: Role;
+    selected?: string | null;
+    role?: Partial<Role>;
+}
+
+const GuestPermissionsTree = ({license, onToggle, readOnly, scope, selectRow, parentRole, selected, role = {permissions: []}}: Props) => {
+    const [editTimeLimitModalIsVisible, setEditTimeLimitModalIsVisible] = React.useState(false);
+
+    const openPostTimeLimitModal = useCallback(() => {
+        setEditTimeLimitModalIsVisible(true);
+    }, []);
+    const closePostTimeLimitModal = useCallback(() => {
+        setEditTimeLimitModalIsVisible(false);
+    }, []);
+    const toggleGroup = useCallback((ids: string[]) => {
+        if (readOnly) {
+            return;
+        }
+        onToggle(role.name!, ids);
+    }, [onToggle, readOnly, role.name]);
+
+    const ADDITIONAL_VALUES = useMemo(() => {
+        return {
+            guest_edit_post: {
+                editTimeLimitButton: (
+                    <EditPostTimeLimitButton
+                        onClick={openPostTimeLimitModal}
+                        isDisabled={readOnly}
+                    />
+                ),
+            },
+        };
+    }, [openPostTimeLimitModal, readOnly]);
+
+    let permissions = useMemo(() => {
+        return [
+            Permissions.CREATE_PRIVATE_CHANNEL,
+            Permissions.EDIT_POST,
+            Permissions.DELETE_POST,
+            Permissions.CREATE_POST,
+            {
+                id: 'guest_reactions',
+                combined: true,
+                permissions: [
+                    Permissions.ADD_REACTION,
+                    Permissions.REMOVE_REACTION,
+                ],
+            },
+            Permissions.USE_CHANNEL_MENTIONS,
+        ];
+    }, []);
+
+    if (license && license.IsLicensed === 'true' && license.LDAPGroups === 'true') {
+        permissions.push(Permissions.USE_GROUP_MENTIONS);
+    }
+    permissions = permissions.map((permission) => {
+        if (typeof (permission) === 'string') {
+            return {
+                id: `guest_${permission}`,
+                combined: true,
+                permissions: [permission],
+            };
+        }
+        return permission;
+    });
```

Comment:
While checking your suggestion @hmhealey , I realized I missed something and I have a doubt here. 
Previously, this was within constructor. With the change the functional component, I just moved out of constructor. 

However, when I click to change any setting, `permission` is recreated with additional item added (please see screen recording below)

What would be the best way to handle this?

https://user-images.githubusercontent.com/37421564/161628390-ea5a3e10-7789-4dc7-97ef-02ac47fc03cf.mov

Reply:
Maybe move all that logic into a function, change `permissions` to be state and then set `permissions` default state using that new function you created? The way it is currently implemented, this code will run every render which is different to what happened before

Reply:
Thanks for the guidance. I tried to use state but `permission` is never updated so `setState` is never used. For this reason, I believe `useRef` is more appropriate here. I updated the code accordingly

Reply:
I'm not sure if `useRef` would work there since I don't think they'd be updated if `license` where to change. I'd think that if you passed `license` and anything else used in `setDefaultPermission` into a `useMemo` like
```javascript
const permissions = useMemo(setDefaultPermission, [license]);
```
then it should re-use the same array unless those values change. I think `useMemo` is basically the same as `useRef` in this case except it automatically updates it when license changes.

Also, `setDefaultPermission` would probably need a different name if we did that or we could go back to just using an arrow function inside the `useMemo`, but that's shouldn't have any effect on the functionality

Reply:
Yes, you are right. I forgot `license` is being used inside `setDefaultPermission`. 
As you suggested, I also changed `setDefaultPermission` name to `setPermissions`